### PR TITLE
Improves Mapping generation and annotations for Entity fields + Bugfixing

### PR DESCRIPTION
**Pull request by D-02022020**
## Annotations

Before | After
-------|------
`@IndexMode(includeInAll = false)` | `@IndexMode(includeInAll = ESOption.FALSE)`
`@IndexMode(normEnabled = false)` | `@IndexMode(normsEnabled = ESOption.FALSE)`
`@IndexMode(indexOptions = IndexMode.INDEX_OPTION_FREQS)` | `@Analyzed(indexOptions = Analyzed.INDEX_OPTION_FREQS)`
`@IndexMode(analyzer = IndexMode.ANALYZER_WHITESPACE)` | `@Analyzed(analyzer = Analyzed.ANALYZER_WHITESPACE)`
`@IndexMode(indexMode = IndexMode.MODE_NO)` | `@IndexMode(indexed = ESOption.FALSE)`
`@IndexMode(indexMode = IndexMode.MODE_NOT_ANALYZED)` | *is the default value*
`@IndexMode(indexMode = IndexMode.MODE_ANALYZE)` | `@Analyzed`
`@Stored` | `@IndexMode(stored = ESOption.TRUE)`

`@Analyzed` makes only sense for fields of type `String`, `List<String>` or `AutoComplete`. 

`@IndexMode.normsEnabled` makes only sense for fields of type `String` or `List<String>`.

`@IndexMode.docValues` makes no sense for fields of type `String` or `List<String>` that are analyzed.

## Default values (almost the same as before):
Option | Value
-------|------
**include_in_all** | `true` for `String` and `List<String>`, `false` otherwise
**norms_enabled** | `false`
**indexed** | `true` (formerly `MODE_NOT_ANALYZED` which means the same)
**stored** | `false` (formerly `true` due to a bug)
**analyzed** | disabled (formerly `MODE_NOT_ANALYZED` which means the same)
**analyzer** | `whitespace` for `AutoComplete`, elasticsearch default (global configured analyzer) otherwise
**index_options** | elasticsearch default (`positions` for analyzed fields, and to `docs` for not_analyzed fields)
**doc_values** (NEW!) | `false` if **indexed** is `false`, elasticsearch default (`true`) otherwise

## The `ESOption` enum

The purpose of the `ESOption` enum is to handle the appliance of default values. When using primitive `boolean`s as values in the annotations' fields, one could not distinguish between purposely set values and annotation default values. This is because an annotation always has values for all properties it declares, not only those that were set in the usage. Furthermore, `null` is not allowed as default value in an annotation. This wouldn't be a problem, if not some subclasses of `Property` wanted to use differing default values (for example for `include_in_all`).

To solve this dilemma, we use `ESOption`, which has 4 possible values instead of 2.
- `TRUE`: use `true`
- `FALSE`: use `false`
- `ES_DEFAULT`: use the default value of elasticsearch by not sending the option at all
- `DEFAULT`: use the default value of sirius-search as defined in the specific subclass of `Property` (may be one of the 3 above)

### CODE REVIEW

**Comment by Tobsucht** [comment]
norms_enabled | elasticsearch default (true for analyzed fields, false otherwise)
--> würde ich default auf false setzen? @andyHa  
das wird eigentlich fast nie gebraucht und braucht unnötig speicher

File: src/main/java/sirius/search/annotations/ListType.java [hunk 1]
```diff
@@ -29,6 +29,4 @@
      * @return the element class of the annotated list field
      */
     Class<?> value();
-
-    boolean nested() default true;
```
Comment (Tobsucht) [inline src/main/java/sirius/search/annotations/ListType.java]: Why remove the control to use nested/object?

File: src/main/java/sirius/search/properties/ObjectListProperty.java [hunk 2]
```diff
@@ -58,6 +57,31 @@ public Property create(Field field) {
...
+    }
+
+    @Override
+    public boolean acceptsSetter() {
+        return false;
+    }
+
+    @Override
+    protected String getMappingType() {
+        return "nested";
```
Comment (Tobsucht) [inline src/main/java/sirius/search/properties/ObjectListProperty.java]: s.o.

File: src/main/java/sirius/search/properties/Property.java [hunk 3]
```diff
@@ -10,67 +10,141 @@
...
 import sirius.web.security.UserContext;
 
 import java.io.IOException;
+import java.lang.annotation.Annotation;
 import java.lang.reflect.Field;
+import java.lang.reflect.Modifier;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import static sirius.search.properties.ESOption.*;
```
Comment (Tobsucht) [inline src/main/java/sirius/search/properties/Property.java]: remove static import

File: src/main/java/sirius/search/annotations/ListType.java (hunk 1 above)
Comment (D3nnisH) [inline src/main/java/sirius/search/annotations/ListType.java]: moved to `@MapType`

File: src/main/java/sirius/search/properties/Property.java [hunk 4]
```diff
@@ -250,22 +450,46 @@ protected Object transformFromRequest(String name, WebContext ctx) {
...
+            }
+        } else {
+            for (Field innerField : nestedClass.getDeclaredFields()) {
+                if (!innerField.isAnnotationPresent(Transient.class) && !Modifier.isStatic(innerField.getModifiers())) {
+                    for (PropertyFactory f : Injector.context().getPartCollection(PropertyFactory.class)) {
+                        if (f.accepts(innerField)) {
+                            Property p = f.create(innerField);
+                            p.setInnerProperty(true);
+                            p.createMapping(builder);
+                            continue;
```
Comment (Tobsucht) [inline src/main/java/sirius/search/properties/Property.java]: remove continue;

Comment (D-02022020) [reply src/main/java/sirius/search/properties/Property.java]: should have been `break;` 😅

File: src/main/java/sirius/search/annotations/ListType.java [hunk 5]
```diff
@@ -30,5 +30,10 @@
      */
     Class<?> value();
 
-    boolean nested() default true;
+    /**
+     * Determines the mapping type in elasticsearch. Ignored if {@link #value()} == {@link String String.class}
+     *
+     * @return whether to use "nested" or "object" as mapping type
+     */
+    boolean nested() default false;
```
Comment (Tobsucht) [inline src/main/java/sirius/search/annotations/ListType.java]: true to be backward compatible?

**Review by meg1502 — APPROVED** [review]

**Pull request merged**